File: frontend/src/backend/postService.js
import {
    fetchConfig,
    appFetch,
} from "./appFetch";

export const createPost = (post, onSuccess, onErrors) => {
    appFetch(
        "/posts/",
        fetchConfig("POST", post),
        onSuccess,
        onErrors
    );
};

export const deletePost = (post, onSuccess, onErrors) => {
    appFetch(
        `/posts/${post.id}`,
        fetchConfig("DELETE", post),
        onSuccess,
        onErrors
    );
};

export const getAllCategories = (onSuccess) => {
    appFetch(
        "/posts/categories",
        fetchConfig("GET"),
        onSuccess
    );
};

export const getUserPosts = ({ page }, onSuccess) => {
    appFetch(
        `/posts/user?page=${page}`,
        fetchConfig("GET"),
        onSuccess
    );
};

export const uploadPostImage = (post, file, onSuccess) =>
    appFetch(`/posts/addImage/${post}`, fetchConfig("PUT", file), onSuccess);

export const getPostByPostId = (id, onSuccess) =>
    appFetch(`/posts/${id}`, fetchConfig("GET"), onSuccess);


export const modifyPost = (id, post, onSuccess, onErrors) => {
    appFetch(
        `/posts/${id}`,
        fetchConfig("PUT", post),
        onSuccess,
        onErrors
    );
};

export const getPosts = ({ title, categoryId, page, size, criteria, order }, onSuccess, onErrors) => {
    let url = `/posts/?page=${page}&size=${size}`;

    if (title) {
        url += `&title=${title}`;
    }

    if (categoryId) {
        url += `&categoryId=${categoryId}`;
    }

    if (criteria) {
        url += `&criteria=${criteria}`;
    }

    if (order !== undefined) {
        url += `&order=${order}`;
    }

    appFetch(
        url,
        fetchConfig("GET"),
        onSuccess,
        onErrors
    );
};

export const newPosts = (timestamp, onSuccess) => {
    appFetch(
        `/posts/new?date=${timestamp}`,
        fetchConfig("GET"),
        onSuccess
    );
};

export const getComments = ({ postId, page }, onSuccess, onErrors) => {
    appFetch(
        `/posts/${postId}/comments?page=${page}`,
        fetchConfig("GET"),
        onSuccess,
        onErrors
    );
};

export const createComment = (postId, comment, onSuccess) => {
    appFetch(
        `/posts/${postId}/comment`,
        fetchConfig("POST", comment),
        onSuccess,
    );
};

export const createAnswer = (id, comment, onSuccess) => {
    appFetch(
        `/posts/comment/${id}/answer`,
        fetchConfig("POST", comment),
        onSuccess,
    );
};

export const deleteComment = (comment, onSuccess, onErrors) => {
    appFetch(
        `/posts/comment/${comment}`,
        fetchConfig("DELETE", comment),
        onSuccess,
        onErrors
    );
};

export const modifyComment = (id, comment, onSuccess) => {
    appFetch(
        `/posts/comment/${id}`,
        fetchConfig("PUT", comment),
        onSuccess,
    );
};

File: overcut/src/main/java/overcut/model/services/PostService.java
package overcut.model.services;

import overcut.model.common.exceptions.InstanceNotFoundException;
import overcut.model.entities.Category;
import overcut.model.entities.Post;
import overcut.model.entities.PostSection;
import overcut.model.entities.User;
import overcut.model.services.exceptions.PermissionException;
import overcut.model.services.exceptions.PostException;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.List;

/**
 * The Interface PostService.
 */
public interface PostService {
    /**
     * Delete a post.
     *
     * @param userId the user id
     * @param postId the post id
     * @throws InstanceNotFoundException the instance not found exception
     * @throws PermissionException       the permission exception
     */
    void deletePost(Long userId, Long postId) throws InstanceNotFoundException, PermissionException;

    /**
     * Get posts.
     *
     * @param title      the title
     * @param categoryId the category id
     * @param page       the page
     * @param size       the size
     * @return the list of posts
     */
   Block<Post> getPosts(String title, Long categoryId, int page, int size, Short criteria, boolean order);

    /**
     * Visualize all the user posts.
     *
     * @param userId the user id
     * @param page   the page
     * @param size   the size
     * @return the list with all the user posts
     * @throws InstanceNotFoundException the instance not found exception
     */
    Block<Post> visualizeAllUserPosts(Long userId, int page, int size) throws InstanceNotFoundException;

    /**
     * Create post.
     *
     * @param title            the title
     * @param subtitle         the subtitle, short description of the article
     * @param userId           the user id
     * @param article          the article, the content news
     * @param categoryId        the category id
     * @return the post
     * @throws InstanceNotFoundException the instance not found exception
     * @throws PostException             the post exception
     */
    Post createPost(String title, String subtitle, String article, Long userId, Long categoryId, byte[] image) throws InstanceNotFoundException, PostException;

    /**
     * Modify post.
     * @param postId           the post id
     * @param title            the title
     * @param subtitle         the subtitle, short description of the article
     * @param userId           the user id
     * @param article          the article, the content news
     * @param categoryId       the category id
     * @return the post
     * @throws InstanceNotFoundException the instance not found exception
     * @throws PostException             the post exception
     */
    Post modifyPost(Long postId, String title, String subtitle, String article, Long userId, Long categoryId, byte[] image)
            throws InstanceNotFoundException, Exception, PostException;


 /**
     * Add image.
     *
     * @param postId the post id
     * @param image  the image
     * @return the post
     * @throws InstanceNotFoundException the instance not found exception
     * @throws IOException               the IO exception
     */
    Post addImage(Long postId, MultipartFile image) throws InstanceNotFoundException, IOException;

    /**
     * Get all categories.
     *
     * @return the list with all categories.
     */
    List<Category> getAllCategories();

    /**
     * Get post details.
     *
     * @param postId the post id
     * @return the post
     * @throws InstanceNotFoundException the instance not found exception
     */
    PostDetails getPostDetails(Long userId, Long postId) throws InstanceNotFoundException;

    boolean newPosts(LocalDateTime referenceDate);


    User getUserPost(Long postId) throws InstanceNotFoundException;

    // Añadir secciones enriquecidas a un post
    void addPostSections(Long postId, List<PostSection> sections) throws InstanceNotFoundException;

    // Obtener secciones enriquecidas de un post
    List<PostSection> getPostSections(Long postId) throws InstanceNotFoundException;

    // Eliminar todas las secciones de un post (por si se quiere regenerar)
    void deletePostSections(Long postId) throws InstanceNotFoundException;

}
